File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples
  ( sumOfMultiples
  ) where

import qualified Data.List as List

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit =
  let validFactors = List.filter (0 /=) factors
   in List.sum $
      List.filter
        (\x -> List.any ((0 ==) . rem x) validFactors)
        [1 .. limit - 1]

File: elixir/sum-of-multiples/lib/sum_of_multiples.ex
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    is_evenly_divisible = fn x ->
      Enum.any?(factors, &(Integer.mod(x, &1) == 0))
    end

    2..(limit - 1)
    |> Enum.to_list()
    |> Enum.filter(is_evenly_divisible)
    |> Enum.sum()
  end
end
